storage/bootstrap/bootstrapper/peers/source.go

```diff
@@ -654,6 +671,90 @@ func (s *peersSource) readBlockMetadataAndIndex(
 	return true, nil
 }
 
+func (s *peersSource) peerAvailability(
+	nsMetadata namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+) result.ShardTimeRanges {
+	var (
+		self                    = s.opts.AdminClient().Options().(client.AdminOptions).Origin().ID()
+		peerAvailabilityByShard = map[uint32]*shardPeerAvailability{}
+	)
+
+	for shardID := range shardsTimeRanges {
+		shardPeers, ok := peerAvailabilityByShard[shardID]
+		if !ok {
+			shardPeers = &shardPeerAvailability{}
+			peerAvailabilityByShard[shardID] = shardPeers
+		}
+		hostShardStates, ok := s.initialTopologyState.shardStates[shardID]
+		if !ok {
+			// This shard was not part of the topology when the bootstrapping
+			// process began.
+			continue
+		}
+
+		shardPeers.numPeers = len(hostShardStates)
+		for _, hostShardState := range hostShardStates {
+			if hostShardState.host.ID() == self {
+				// Don't take self into account
+				continue
+			}
+
+			shardState := hostShardState.shardState
+
+			switch shardState {
+			// Skip cases - We cannot bootstrap from this host
+			case shard.Initializing:
+				// Don't want to peer bootstrap from a node that has not yet completely
+				// taken ownership of the shard.
+			case shard.Unknown:
+
+				// Success cases - We can bootstrap from this host, which is enough to
+				// mark this shard as bootstrappable.
+			case shard.Leaving:
+				fallthrough
+			case shard.Available:
+				shardPeers.numAvailablePeers++
+				break
```

Comment:
I don't think you need an explicit break here, should break correctly without falling through unless explicitly falling through.